### ISSUE TITLE
Many std iterators do not implement Send/Sync

### ISSUE THREAD

**Issue by huonw** [rust-lang/rust#22709]
Many iterators use `*` internally, for performance. This means they don't automatically implement `Send` or `Sync`, but many/most can.

Based on the regexp, `\bIterator for`, the following (non-compiler/non-test) files have `Iterator` implementations. We should look through them to find raw pointers etc. and ensure that they implement `Send`/`Sync` if there are such pointers. We can do it progressively, comment below if you feel like taking some files:

- [ ] liballoc/boxed.rs
- [x] libcollections @edwardw #22715
  - /binary_heap.rs 
  - /bit.rs
  - /btree/map.rs
  - /btree/node.rs
  - /btree/set.rs
  - /enum_set.rs
  - /linked_list.rs
  - /slice.rs
  - /str.rs
  - /vec.rs
  - /vec_deque.rs
  - /vec_map.rs
- [x] libcore/char.rs @Nashenas88 #26041
- [x] libcore/iter.rs @Nashenas88 #26126
- [ ] libcore/option.rs
- [ ] libcore/result.rs
- [ ] libcore/slice.rs
- [ ] libcore/str/mod.rs
- [ ] libfmt_macros/lib.rs
- [ ] librand/lib.rs
- [ ] libserialize/json.rs
- [ ] libstd/ascii.rs
- [x] libstd/collections/hash/map.rs @GuillaumeGomez #27615
- [x] libstd/collections/hash/set.rs @GuillaumeGomez #27615
- [x] libstd/collections/hash/table.rs @GuillaumeGomez #27615
- [ ] libstd/env.rs
- [ ] libstd/fs.rs
- [ ] libstd/io/mod.rs
- [ ] libstd/net/mod.rs
- [ ] libstd/net/tcp.rs
- [ ] libstd/path.rs
- [ ] libstd/sync/mpsc/mod.rs
- [ ] libstd/sync/mpsc/select.rs
- [ ] libstd/sys/common/net.rs
- [ ] libstd/sys/common/wtf8.rs
- [ ] libstd/sys/unix/fs2.rs
- [ ] libstd/sys/unix/os.rs
- [ ] libstd/sys/windows/fs2.rs
- [ ] libstd/sys/windows/os.rs
- [ ] libunicode/u_str.rs

The `&mut self` of `Iterator` means that almost all iterators will be safe to implement `Send` and `Sync`.

**Comment by edwardw** [comment]
I guess ` libcollections` is mine.

**Comment by mund** [comment]
I would like to take on some files, but slightly unclear on what to do. For instance, in liballoc/boxed.rs:

    #[stable(feature = "rust1", since = "1.0.0")]
    impl<I: Iterator + ?Sized> Iterator for Box<I> {
        type Item = I::Item;
        fn next(&mut self) -> Option<I::Item> { (**self).next() }
        fn size_hint(&self) -> (usize, Option<usize>) { (**self).size_hint() }
    }

Where would I implement Send/Sync? I'm looking at #22715 for examples/guidance but not getting much out of it. Once I can do that single file, it would be my pleasure to work on the libcore files as well.

**Comment by huonw** [comment]
Thanks @mund.

That impl is saying that if a `Box` contains an iterator, then the whole pointer can be treated as an iterator, i.e. just a proxy implementation to allow `Box`s to be thrown around more freely. `Box` already implements `Send` and `Sync` so this case is handled.

The "interesting" ones (for this bug) are when there's a `struct` that's defined purely to be an iterator, e.g. [the `Iter` struct](https://github.com/rust-lang/rust/blob/master/src/libcore/slice.rs#L737-L741) in libcore/slice.rs. In particular, that struct contains raw pointers (`*const`'s), and so does not automatically implement `Send` or `Sync`.

However, it is threadsafe (with the appropriate bounds on the `T` type) so implementing them manually is correct!

For that specific type:

```rust
unsafe impl<'a, T: Sync> Sync for Iter<'a, T> {}
unsafe impl<'a, T: Sync> Send for Iter<'a, T> {}
```

The appropriate bounds to use depends on the element type that the iterator yields, iterators (for the most part) can be regarded as having the same send/sync-ness as the element type, meaning if the element type of an iterator `Foo<T>` is:

- `&T`, the appropriate bounds are probably `T: Sync` for both `Sync` and `Send`,
- `&mut T`, the appropriate bounds are probably `T: Sync` for `Sync` and `T: Send` for `Send`
- `T`, the appropriate bounds are probably `T: Sync` for `Sync` and `T: Send` for `Send` (same as `&mut T`)

In summary, it is likely only the iterator structs/enums that contain raw pointers that need to be considered for this bug.

**Comment by edwardw** [comment]
@mund, now #22715 has been merged with an extensive test. You are welcome to use it as a template to work on other libs.

**Comment by mund** [comment]
Thanks Edward! I'll work on this tonight.

**Comment by djmally** [comment]
Is there still more work that needs to be done here? I'd like to help contribute to this, but it's not quite clear what's been done yet.

**Comment by Nashenas88** [comment]
@djmally Take a look at @edwardw's pull request. It looks like that's the only one that's been done so far.

**Comment by edwardw** [comment]
@djmally, what @Nashenas88 said. The `libcollections` is the only one that has been done. You are welcome to use my PR as a template to work on other libraries.

**Comment by Nashenas88** [comment]
`libcore/char.rs` is only implementing `Iterator`s where `Item` is `char`. I guess this file doesn't apply for this bug? Would it get `Sync` and `Send` automatically?

**Comment by Nashenas88** [comment]
I've run `grep -E -e '\bIterator[[:space:]]+for' * -RI | sed -E 's/^(.*\.rs:).*Iterator for ([^{]+)[[:space:]]*.*$/\1\2/'` (note, this is mac/bsd syntax) to get the list of types that `Iterator` is being implemented for (without any trailing whitespace). The output format is `filename:type`.

**Comment by huonw** [comment]
> I guess this file doesn't apply for this bug? Would it get Sync and Send automatically?

It does: whether or not a type is `Sync`/`Send` by default depends on its internals, e.g. if it has a raw pointer it won't implement either trait by default.

**Comment by edwardw** [comment]
@Nashenas88, the most relevant part of my original PR is the macros for testing `Sync / Send` after you obtain the list of all types which implement `Iterator`.

And BTW, I think the `E-easy` tag is a bit misleading. This ticket is quite subtle actually.

**Comment by Nashenas88** [comment]
@huonw, to clarify, libcore/char.rs has two implementers of `Iterator`, neither of which is using raw pointers. The only thing that stands out is an unsafe transmute from `i32` to `char`.

@edwardw, should I add tests for this file to verify that it implements `Sync / Send`, even if it already implements them by default (more of a regression test)?

**Comment by edwardw** [comment]
@Nashenas88, yes, that's the point of this ticket I believe. The tests you are adding will verify that iterators do implement `Sync / Send` and safeguard against any future regression.

**Comment by Nashenas88** [comment]
The test for char is merged. Could the list be updated? Also, am I missing anything in the test?

**Comment by bluss** [comment]
Thanks @Nashenas88! I updated the list.

**Comment by Nashenas88** [comment]
Woot! My first **real** contribution. I'm on the 1.0 contributors list because of a tiny typo fix to the documentation :smile:. @huonw, could I own the rest of libcore? They don't seem complicated. I'll probably get started on them on Tuesday at the latest.

**Comment by Veedrac** [comment]
I did a quick pass to knock things off the list.

These *might* need adding `Send` or `Sync`

* `libfmt_macros/lib.rs`
* `libstd/collections/hash/map.rs` (see: `hash/table.rs`)
* `libstd/collections/hash/set.rs` (see: `hash/table.rs`)
* `libstd/collections/hash/table.rs`
* `libstd/sync/mpsc/select.rs`
* `libstd/sys/common/net2.rs`
* `libstd/sys/windows/os.rs` (`LPWCH`)

These are fine

* `liballoc/boxed.rs`
* `libcore/option.rs`
* `libcore/result.rs`
* `librand/lib.rs`
* `libserialize/json.rs`
* `libstd/ascii.rs`
* `libstd/fs.rs`
* `libstd/io/mod.rs`
* `libstd/net/mod.rs`
* `libstd/net/tcp.rs`
* `libstd/path.rs`
* `libstd/sync/mpsc/mod.rs`
* `libstd/sys/common/wtf8.rs`
* `libstd/sys/unix/fs.rs` (prev `fs2`, I assume)
* `libstd/sys/windows/fs.rs` (prev `fs2`, I assume)
* `libunicode/u_str.rs`

Miscellania

* `libcore/iter.rs` (#26126)
* `libcore/slice.rs` → Does fancy things with macros. Needs more eye-time.
* `libcore/str/mod.rs` → Does fancy things with macros. Needs more eye-time.
* `libstd/env.rs` → Probably fine, but needs double-checking.
* `libstd/sys/unix/os.rs` → `_dont_send_or_sync_me: *mut (),` should probably be an opt-out.
* `libstd/old_io/*` is dead, methinks.

**Comment by Veedrac** [comment]
I'm pretty sure `libstd/sys/unix/os.rs`'s iterators are safe since they're over owned vectors, but they explicitly opt out.

This makes me think they are opting out to match `libstd/sys/windows/os.rs`, but those ones seem like they could well be `Send` and `Sync` too.

I'll put up a pull request making these changes, but I imagine those warnings were put there for a reason. Pinging @alexcrichton since he seems to have written this.

**Comment by alexcrichton** [comment]
Yeah you're correct that the Unix iterators are fine to be Send/Sync, they only currently opt-out because the Windows ones are not Send/Sync (due to their storage of raw pointers). I haven't done the necessary leg work to ensure that the Windows implementations are indeed safe to send across threads, so I figured I'd just make the Unix ones not Send/Sync (e.g. the more conservative choice).

I'd probably avoid adding Send/Sync to those iterators for now as needing an iterator to be Send/Sync is quite rare, and even rarer for the two iterators in question. Plus, it's always possible to make these Send/Sync backwards compatibly!

**Comment by Veedrac** [comment]
Ah, great. I think all you'd have to do is mark them as `Send`/`Sync` since the memory they use is thread-safe, but I agree there's no rush. (If they're not thread-safe we're doomed anyway!)

**Comment by GuillaumeGomez** [comment]
Is there still some work here? I'm not sure the first post is up-to-date.

**Comment by Veedrac** [comment]
@GuillaumeGomez AFAIK, [my later comment](https://github.com/rust-lang/rust/issues/22709#issuecomment-110935709) is still up to date.

**Comment by GuillaumeGomez** [comment]
Perfect, I'll take a look then!

**Comment by GuillaumeGomez** [comment]
I'm working on `libstd/collections`.

**Comment by GuillaumeGomez** [comment]
I take `libstd/sys/common/net2.rs`.

**Comment by brson** [comment]
This looks mostly settled, but hard to tell the state of things. Presumably if there are more problems related to this they will show up as bug reports someday. Closing. Feel free to reopen if you disagree.
